### ISSUE TITLE
r10k with each (internal) module in its own repo?

### ISSUE THREAD

**Issue by antaflos** [puppetlabs/r10k#19]
We use about 100 Puppet modules, most of them written by us, some of them from the Forge or directly from Github. Our own modules each live in their own Git repository, as in `git@ops.example.com/puppet-tomcat`, `git@ops.example.com/puppet-chrony` and so on. Currently we don't even use environments, the modules are cloned directly into `/etc/puppet/modules`, either from the Forge, Github or our own Git server. This sucks, obviously. 

So my question: is r10k useful for us? We don't have a single Git repo containing all modules (no org-shared-repo as the example r10k.yaml file suggests), so I don't understand if r10k can cope with this kind of setup. Seems like it easily could, but I am not sure how to start deploying it. Any hints?

**Comment by adrienthebo** [comment]
If I understand this correctly, you probably want something like what a Puppetfile does. For example:

``` ruby
# installs from git
mod 'boolean', :git => 'git://github.com/adrienthebo/puppet-boolean'

# installs from the puppet forge
mod 'cloudsmith/demoapp'
mod 'cloudsmith/demoplatform', '0.1.5'
```

This basic Puppetfile will install the `boolean`, `demoapp`, and `demoplatform modules`. The current behavior is that r10k creates a directory called `modules/` in the same directory as the Puppetfile, and installs all of the modules there. In this case you would create `/etc/puppet/Puppetfile` and run something like `cd /etc/puppet; r10k puppetfile install` and all of the modules specified in the Puppetfile would be updated as needed. The Puppetfile doesn't need to be version controlled, and doesn't need to be contained.

Is this the behavior you're looking for?

**Comment by adrienthebo** [comment]
Whoa, sorry, didn't mean to close this! Anyhow, does this look like what you need?

**Comment by antaflos** [comment]
Thanks for the reply and information! It'll certainly prove useful.

In the end, I want to be able to use r10k and dynamic Puppet environments in the way you described in various blog entries. But since we don't have a single Git repo containing all modules (instead we have one repo per module) I am not sure if we can even implement a proper dynamic environment workflow. r10k seems to expect a repo like `git@github.com:my-org/org-shared-modules` where multiple modules live in a single repo. We cannot provide such a repo. Can we still use r10k and dynamic environments properly?

**Comment by justenwalker** [comment]
One strength of using git is that all repositories are local.  You could make a local repository on the server which just has Puppetfile, as described by Adrien, and instruct r10k.yaml to use it.  This way you get the benefit of version-controlling your Puppetfile but don't need to set up any additional infrastructure to support it. The default environment that puppet uses is 'production' - so as long as your have a `production` branch in the local repository it should still work.

I'll concede that this is a bit sloppy, and the feature enhancement is still valuable - but in it's absence I would try this:

``` bash
$ git init repository
Initialized empty Git repository in /path/to/local/git/repository
$ cd repository/
# Edit Puppetfile and populate with all your git repositories
$ git add Puppetfile 
$ git commit -m 'Added Puppetfile'
$ git branch production; git checkout production
```

Then reference the local repository in `/etc/r10k.yaml`

``` yaml
---
:cachedir: '/var/cache/r10k'
:sources:
  :local:
    remote: '/path/to/local/git/repository'
    basedir: '/etc/puppet/environments'

:purgedirs:
  - '/etc/puppet/environments'
```

and try `r10k synchronize`

**Comment by antaflos** [comment]
@justenwalker thanks, that works fine, it seems! At least now we can combine all our repos (modules) into one directory and keep them updated.

But how would a workflow based on dynamic Puppet environments work here? Say I want to implement a new feature in our `puppet-tomcat` module. I'd create a new branch `my_new_feature` in the `puppet-tomcat` repo, implement the feature and push that branch to our main Git repository server. But before merging it into the master branch of the `puppet-tomcat` repo, I'd like to test it.

What would I do? I suspect I'd checkout a new branch in `/path/to/local/git/repository`, possibly calling it `test_my_new_feature` and then edit the Puppetfile, specifying `:ref => 'my_new_feature` for the `puppet-tomcat` module, and then run `r10k synchronize`?

Is that a proper approach? Maybe I don't understand what to expect but it seems a bit tedious to me.

**Comment by justenwalker** [comment]
@antaflos 

**Note**: I think I may be abusing the issue comments as if it were a Stack Overflow question. It may be valuable to pose this question more generally through official channels, as this work-flow is not specific to r10k, and this medium effectively limits your audience.

The above notwithstanding; In the case you are describing, it would be much more sane to make a central repository to store your Puppetfile instead of the local one that I suggested before, even if it only contains that one Puppetfile. Otherwise, you'll have to log into the puppet master every time and fiddle around committing and merging branches. I think this is where the tedium you speak of may be.

Let's say you have 2 branches: `production` and `testing`.  `production` you use on your main servers. and `testing` you use for puppet testing like for tomcat.

You could have a more persistent branch on each module that you push to in order to test on the 'testing' environment - and always have modules referencing it instead. In which case you wont have to fiddle with the Puppetfile as often. Just push your module's feature branch to the testing branch when you are ready to test it.

Example:

**Puppetfile `testing` Branch**

```
# Tomcat module
mod 'tomcat', 
  :git => 'git://git.ourcompany.com/puppet-modules/tomcat.git',
  :ref => 'testing'

# Repeat for other modules...
mod 'jenkins', 
  :git => 'git://git.ourcompany.com/puppet-modules/jenkins.git',
  :ref => 'testing'

...

```

**Puppetfile `production` Branch**

```
# Tomcat module
mod 'tomcat', 
  :git => 'git://git.ourcompany.com/puppet-modules/tomcat.git',
  :ref => 'master'

# Repeat for other modules...
mod 'jenkins', 
  :git => 'git://git.ourcompany.com/puppet-modules/jenkins.git',
  :ref => 'master'

...

```

When r10k synchronizes your repository on the puppet master, it will create 2 folders:
- `/etc/puppet/environments/testing`
- `/etc/puppet/environments/production`

Most servers will be requesting modules from the `production` folder since:
1. Your puppet-master's `modulepath` will be set to something like: `modulepath=/etc/puppet/environments/$environment/modules`
2. The majority of your servers will be using the default `environment = production` setting in puppet.conf

However, your testing server(s) will request the `testing` folder since it will have `environment = testing` in `/etc/puppet/puppet.conf` and it will get the new tomcat module.

Finally, once you have accepted this feature, you can merge `my_new_feature` into `master` and now production servers will have it.

**Comment by antaflos** [comment]
@justenwalker thank you very much for that detailed description! We have now implemented a workflow more or less exactly like that and so far it works reasonably well. Of course it is not as dynamic as the workflow described in https://puppetlabs.com/blog/git-workflow-and-puppet-environments/ (because the environment branches are fixed, in this case to `production` and `testing`) but it is a huge step forward from a static module tree with no environments at all.

Unfortunately, I am not yet quite happy, because the workflow now looks like this:
1. Set up Puppetfile in `git@git.example.com:puppetmodules` for both `production` and `testing` environments as described
2. Run `r10k synchronize`
3. The Tomcat module now needs a new feature
4. Clone `git@git.example.com:puppet-tomcat` to my workstation
5. Create a feature branch or work in `puppet-tomcat`'s testing branch
6. Hack away on the new feature
7. Commit the new feature to the feature or testing branch
8. Push the local branch to the remote testing branch
9. Run `r10k synchronize` which updates all modules and gets the newest commits from the remote puppet-tomcat module repo (testing branch) into `/etc/puppet/environments/testing/modules/tomcat`
10. Run `puppet agent -t --environment testing` on a testing machine
11. The testing machine can now make use of the new feature implemented in the puppet-tomcat module

The problem here is, obviously, that it takes about 8 steps for me to know whether my new feature in the puppet-tomcat module even works as intended. If it doesn't I have to go back to 6 and start over. It also means I have to always push commits to the remote repo, even though those commits might be full of errors and problems. 

It's a bit like not being able to compile a C program locally; instead I have to blindly implement the desired feature locally, commit and push to a remote repo, and only then the program gets compiled. And only then I get to see if my new feature works (or even compiles). 

There has to be a better way, no?

As to abusing the issue comments for general discussion: is there a more proper forum for discussing r10k and workflows for dynamic environmens? The puppet-users list?

**Comment by justenwalker** [comment]
@antaflos It isn't obvious to me that there is an easier way without a build automation tool (such as Jenkins) to reduce this tedium to something more manageable. Unless @adrienthebo has some additional input on this, I see supporting your workflow natively in r10k requiring heavier architectural changes.

I cannot agree with all of the steps you listed as being in scope of the problem:
- 1 and 2 are one time configurations
- 3 through 6 can be ignored as you must do this even without r10k
- 7 & 8 are essentially the same

Perhaps I'm oversimplifying, but as I see it from r10k's perspective there is:
- a short one-time setup for r10k Puppetfile repository
- a somewhat longer one-time build-automation setup in Jenkins (or your CI tool of choice)
- git push origin my_feature:testing
- a build-automation task triggered by a push on a 'testing' branch to synchronize and run puppet on the test environment

From a developer's perspective, the workflow would look like:
- make my feature ... commit to my branch
- `git push my_feature:testing # kicks off the build too`
- Wait for the ci-server to succeed/fail and notify me

Update/Clarification: The build automation step would most likely be per module in your case, or the SCM trigger would have to be set up for each module repository to kick off the more general build/test task.

https://ask.puppetlabs.com/questions/ is the puppetlabs Q&A site.
puppet-users is pretty active - and it persists better than IRC chats. That said I personally prefer IRC. (#puppet on freenode)

**Comment by nanliu** [comment]
@antaflos why not test with something like vagrant? That will give you much shorter feedback loop than doing this via master/agent. Even puppet apply would simplify the workflow over your existing process. It should really boil down to a local save, test, commit loop before you push the changes and fetch on master.

**Comment by antaflos** [comment]
@justenwalker and @nanliu thank you both for your very helpful input. I've started looking into Vagrant for testing modules and manifests, and that seems to be along the lines of what I hope to achieve eventually (getting more "instant" feedback on new features in modules). And once we have Rspec tests in place we can even feed our modules to our Jenkins instance.

See, until I read about r10k we did not even have a proper workflow for our Puppet modules and manifests, and so I've only recently started trying to improve that. Thus my views and thoughts on development/deployment workflows may be naive and/or wrong :) This is of course not specific to r10k and I have probably abused the issue comments and @justenwalker's patience enough for now :) Thanks again!
